### DISCUSSION TITLE
Energy calculation wrong on Nous A1T

### DISCUSSION THREAD

**Discussion by H-Sachse** [arendst/Tasmota#20653] [Support request] [answered]
### PROBLEM DESCRIPTION
The energy calculation on the Nous A1T is off by a fair margin.
e.g. I have a load of 2.3kW (as reported by the A1T for 1hour which should lead to an energy value of 2.3kWh, but the A1T reports 2.9kWh

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [x] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [x] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [x] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [x] Device used (e.g., Sonoff Basic): Nous A1T
- [x] Tasmota binary firmware version number used: 13.3.0
  - [x] Pre-compiled
  - [ ] Self-compiled
- [x] Flashing tools used: OTA
- [x] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:
13:30:46.540 MQT: tasmota/stat/tasmota_158777/RESULT = {"NAME":"NOUS A1T","GPIO":[32,0,0,0,2720,2656,0,0,2624,320,224,0,0,0],"FLAG":0,"BASE":49}
13:30:46.773 MQT: tasmota/stat/tasmota_158777/RESULT = {"Module":{"0":"NOUS A1T"}}
13:30:46.980 MQT: tasmota/stat/tasmota_158777/RESULT = {"GPIO0":{"32":"Button1"},"GPIO1":{"0":"None"},"GPIO2":{"0":"None"},"GPIO3":{"0":"None"},"GPIO4":{"2720":"BL0937 CF"},"GPIO5":{"2656":"HLWBL CF1"},"GPIO9":{"0":"None"},"GPIO10":{"0":"None"},"GPIO12":{"2624":"HLWBL SEL_i"},"GPIO13":{"320":"Led_i1"},"GPIO14":{"224":"Relay1"},"GPIO15":{"0":"None"},"GPIO16":{"0":"None"},"GPIO17":{"0":"None"}}
```
- [x] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:
13:43:22.239 MQT: tasmota/stat/tasmota_158777/RESULT = {"Rule1":{"State":"ON","Once":"ON","StopOnError":"OFF","Length":39,"Free":472,"Rules":"ON Energy#Power < 100 DO SaveData ENDON"}}
```
- [x] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:
13:31:46.899 MQT: tasmota/stat/tasmota_158777/STATUS = {"Status":{"Module":0,"DeviceName":"CarCharger","FriendlyName":["Auto Ladegerät in der Garage"],"Topic":"tasmota_158777","ButtonTopic":"0","Power":1,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0,"StatusRetain":0}}
13:31:46.909 MQT: tasmota/stat/tasmota_158777/STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota/release/tasmota.bin.gz","RestartReason":"Software/System restart","Uptime":"0T00:29:47","StartupUTC":"2024-02-03T12:01:59","Sleep":50,"CfgHolder":4617,"BootCount":10,"BCResetTime":"2024-01-03T15:23:55","SaveCount":198,"SaveAddress":"FA000"}}
13:31:46.917 MQT: tasmota/stat/tasmota_158777/STATUS2 = {"StatusFWR":{"Version":"13.3.0(tasmota)","BuildDateTime":"2023-12-12T14:29:05","Boot":31,"Core":"2_7_4_9","SDK":"2.2.2-dev(38a443e)","CpuFrequency":80,"Hardware":"ESP8266EX","CR":"383/699"}}
13:31:46.926 MQT: tasmota/stat/tasmota_158777/STATUS3 = {"StatusLOG":{"SerialLog":2,"WebLog":2,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["HSMHS8",""],"TelePeriod":60,"Resolution":"558180C0","SetOption":["00008009","2805C80001000600003C5A0A192800000000","00000080","00006000","00004000","00000000"]}}
13:31:46.943 MQT: tasmota/stat/tasmota_158777/STATUS4 = {"StatusMEM":{"ProgramSize":637,"Free":364,"Heap":20,"ProgramFlashSize":1024,"FlashSize":2048,"FlashChipId":"156085","FlashFrequency":40,"FlashMode":"DOUT","Features":["0809","8F9AC787","04368001","000000CF","010013C0","C000F981","00004004","00001000","54000020","00000080","00000000"],"Drivers":"1,2,3,4,5,6,7,8,9,10,12,16,18,19,20,21,22,24,26,27,29,30,35,37,45,62,68","Sensors":"1,2,3,4,5,6","I2CDriver":"7"}}
13:31:46.956 MQT: tasmota/stat/tasmota_158777/STATUS5 = {"StatusNET":{"Hostname":"tasmota-158777-1911","IPAddress":"10.8.10.176","Gateway":"10.8.10.1","Subnetmask":"255.255.255.0","DNSServer1":"10.8.10.1","DNSServer2":"10.8.10.3","Mac":"C8:C9:A3:15:87:77","Webserver":2,"HTTP_API":1,"WifiConfig":4,"WifiPower":17.0}}
13:31:46.967 MQT: tasmota/stat/tasmota_158777/STATUS6 = {"StatusMQT":{"MqttHost":"10.8.100.11","MqttPort":1883,"MqttClientMask":"DVES_%06X","MqttClient":"DVES_158777","MqttUser":"tasmota","MqttCount":1,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30,"SOCKET_TIMEOUT":4}}
13:31:46.980 MQT: tasmota/stat/tasmota_158777/STATUS7 = {"StatusTIM":{"UTC":"2024-02-03T12:31:46","Local":"2024-02-03T13:31:46","StartDST":"2024-03-31T02:00:00","EndDST":"2024-10-27T03:00:00","Timezone":"+01:00","Sunrise":"08:18","Sunset":"17:49"}}
13:31:46.991 MQT: tasmota/stat/tasmota_158777/STATUS9 = {"StatusPTH":{"PowerDelta":[0,0,0],"PowerLow":0,"PowerHigh":0,"VoltageLow":0,"VoltageHigh":0,"CurrentLow":0,"CurrentHigh":0}}
13:31:47.001 MQT: tasmota/stat/tasmota_158777/STATUS10 = {"StatusSNS":{"Time":"2024-02-03T13:31:46","ENERGY":{"TotalStartTime":"2024-01-03T15:23:55","Total":105.581,"Yesterday":13.898,"Today":4.643,"Power":2319,"ApparentPower":2319,"ReactivePower":0,"Factor":1.00,"Voltage":236,"Current":9.830}}}
13:31:47.014 MQT: tasmota/stat/tasmota_158777/STATUS11 = {"StatusSTS":{"Time":"2024-02-03T13:31:47","Uptime":"0T00:29:48","UptimeSec":1788,"Heap":23,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"POWER":"ON","Wifi":{"AP":1,"SSId":"HSMHS8","BSSId":"68:FF:7B:EF:51:8C","Channel":2,"Mode":"11n","RSSI":44,"Signal":-78,"LinkCount":1,"Downtime":"0T00:00:04"}}}
```
- [x] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:
13:35:08.382 MQT: tasmota/tele/tasmota_158777/STATE = {"Time":"2024-02-03T13:35:08","Uptime":"0T00:33:09","UptimeSec":1989,"Heap":23,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"POWER":"ON","Wifi":{"AP":1,"SSId":"HSMHS8","BSSId":"68:FF:7B:EF:51:8C","Channel":2,"Mode":"11n","RSSI":40,"Signal":-80,"LinkCount":1,"Downtime":"0T00:00:04"}}
13:35:08.393 MQT: tasmota/tele/tasmota_158777/SENSOR = {"Time":"2024-02-03T13:35:08","ENERGY":{"TotalStartTime":"2024-01-03T15:23:55","Total":105.748,"Yesterday":13.898,"Today":4.810,"Period":49,"Power":2315,"ApparentPower":2315,"ReactivePower":0,"Factor":1.00,"Voltage":236,"Current":9.830}}
13:35:09.200 HLW: EPC 1734, CFlen 573 usec
13:35:10.201 HLW: EPC 1735, CFlen 573 usec
13:35:11.200 HLW: EPC 1733, CFlen 574 usec
13:35:12.200 HLW: EPC 1733, CFlen 578 usec
13:35:13.200 HLW: EPC 1734, CFlen 577 usec
13:35:14.200 HLW: EPC 1734, CFlen 574 usec
13:35:15.200 HLW: EPC 1734, CFlen 577 usec
13:35:16.200 HLW: EPC 1734, CFlen 578 usec
13:35:17.200 HLW: EPC 1734, CFlen 578 usec
13:35:18.200 HLW: EPC 1734, CFlen 573 usec
13:35:19.200 HLW: EPC 1734, CFlen 574 usec
13:35:20.200 HLW: EPC 1735, CFlen 578 usec
13:35:21.200 HLW: EPC 1733, CFlen 574 usec
13:35:22.200 HLW: EPC 1735, CFlen 578 usec
13:35:22.783 WIF: Checking connection...
13:35:23.200 HLW: EPC 1734, CFlen 574 usec
13:35:24.200 HLW: EPC 1734, CFlen 573 usec
13:35:25.201 HLW: EPC 1735, CFlen 574 usec
13:35:26.200 HLW: EPC 1733, CFlen 584 usec
13:35:27.200 HLW: EPC 1734, CFlen 573 usec
13:35:28.200 HLW: EPC 1734, CFlen 573 usec
13:35:29.200 HLW: EPC 1734, CFlen 578 usec
13:35:30.200 HLW: EPC 1735, CFlen 578 usec
13:35:31.200 HLW: EPC 1734, CFlen 573 usec
13:35:32.200 HLW: EPC 1735, CFlen 577 usec
13:35:33.200 HLW: EPC 1734, CFlen 578 usec
13:35:34.200 HLW: EPC 1734, CFlen 578 usec
13:35:35.200 HLW: EPC 1735, CFlen 578 usec
13:35:36.201 HLW: EPC 1735, CFlen 578 usec
13:35:37.200 HLW: EPC 1734, CFlen 578 usec
13:35:38.200 HLW: EPC 1734, CFlen 573 usec
13:35:39.200 HLW: EPC 1735, CFlen 578 usec
13:35:40.200 HLW: EPC 1734, CFlen 574 usec
13:35:41.199 HLW: EPC 1734, CFlen 573 usec
13:35:42.200 HLW: EPC 1734, CFlen 578 usec
13:35:42.812 WIF: Checking connection...
13:35:43.200 HLW: EPC 1735, CFlen 574 usec
13:35:44.200 HLW: EPC 1734, CFlen 578 usec
13:35:45.200 HLW: EPC 1735, CFlen 578 usec
13:35:46.200 HLW: EPC 1734, CFlen 574 usec
13:35:47.200 HLW: EPC 1734, CFlen 574 usec
13:35:48.200 HLW: EPC 1735, CFlen 578 usec
13:35:49.200 HLW: EPC 1734, CFlen 578 usec
13:35:50.200 HLW: EPC 1734, CFlen 573 usec
13:35:51.200 HLW: EPC 1735, CFlen 570 usec
13:35:52.200 HLW: EPC 1735, CFlen 574 usec
13:35:53.200 HLW: EPC 1734, CFlen 574 usec
13:35:54.200 HLW: EPC 1735, CFlen 573 usec
13:35:55.200 HLW: EPC 1733, CFlen 573 usec
13:35:56.200 HLW: EPC 1734, CFlen 573 usec
13:35:57.200 HLW: EPC 1734, CFlen 578 usec
13:35:58.200 HLW: EPC 1734, CFlen 571 usec
13:35:59.200 HLW: EPC 1735, CFlen 580 usec
13:35:59.374 WIF: Sending Gratuitous ARP
13:36:00.200 HLW: EPC 1735, CFlen 573 usec
13:36:01.200 HLW: EPC 1733, CFlen 577 usec
13:36:02.200 HLW: EPC 1734, CFlen 578 usec
13:36:02.754 WIF: Checking connection...
13:36:03.201 HLW: EPC 1736, CFlen 578 usec
13:36:04.200 HLW: EPC 1731, CFlen 574 usec
13:36:05.200 HLW: EPC 1734, CFlen 573 usec
13:36:06.200 HLW: EPC 1733, CFlen 577 usec
13:36:07.200 HLW: EPC 1733, CFlen 573 usec
13:36:08.200 HLW: EPC 1733, CFlen 578 usec
13:36:08.410 MQT: tasmota/tele/tasmota_158777/STATE = {"Time":"2024-02-03T13:36:08","Uptime":"0T00:34:09","UptimeSec":2049,"Heap":23,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"POWER":"ON","Wifi":{"AP":1,"SSId":"HSMHS8","BSSId":"68:FF:7B:EF:51:8C","Channel":2,"Mode":"11n","RSSI":42,"Signal":-79,"LinkCount":1,"Downtime":"0T00:00:04"}}
13:36:08.425 MQT: tasmota/tele/tasmota_158777/SENSOR = {"Time":"2024-02-03T13:36:08","ENERGY":{"TotalStartTime":"2024-01-03T15:23:55","Total":105.797,"Yesterday":13.898,"Today":4.859,"Period":49,"Power":2323,"ApparentPower":2323,"ReactivePower":0,"Factor":1.00,"Voltage":236,"Current":9.830}}
13:36:09.200 HLW: EPC 1734, CFlen 573 usec
```

### TO REPRODUCE
_Steps to reproduce the behavior:_
Plug in constant load for 1 hour

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._
Energy value reported after one hour equals power of plugged load * 1h, e.g. 2kW * 1h = 2kWh

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by Jason2866** [comment]
Device needs calibration. When done it was not precise enough.

**Reply by H-Sachse** [reply]
I explicitly stated that I was using the Power value as given by Tasmota as a baseline. 
The plug shows 2.3kW with minimal fluctuations for the entire measurement duration (monitored via Web interface and MQTT) and I got 2.9kWh of energy displayed after this hour. What does that have to do with calibration? 
There's an internal calculation difference between 2.3kW * 60 minutes and the actual result.
There is no way to calibrate the energy. I did calibrate Voltage and Current and Voltage * Current = Power. I checked the voltage, current and power values and they are consistent, but the energy value does not match.

**Comment by Jason2866** [comment]
Calibration has influence in every value. Volt, Current and **Power** needs to be set and checked.
https://tasmota.github.io/docs/Power-Monitoring-Calibration/#calibration-procedure
Especially the Nous A1T is very hard to calibrate. I did not got good results without the Fine tuning.
Do the calibration only with a pur resistive load. Bigger load is better. Without having 2 multimeters measuring current and voltage at the same time and comparing the values shown in Web GUI there will no statisfying result.

Your calculation is only correct when it is a 100% resistive load.

**Reply by H-Sachse** [reply]
As I already said (twíce), the values reported by the A1T are in themselves inconsistent!
The A1T reports 2.3kW for 1 hour and then reports 2.9kWh of energy usage. This makes no sense and no calibration will change this discrepancy. I am NOT talking about a discrepancy between real load values and reported values.
Either there's something fishy about how the energy value is accumulated from the power value (I suspect that the integration interval is shorter than expected) or the energy value is calculated from a different (internal) power value which would not make sense.
As can be seen from my log data, the power value is only off by 0.5% from the reported voltage * reported current and not 25% off like the reported energy in relation to the expected value.
Also my logs show the load has a power factor of 1 which means it behaves purely resistive.

**Reply by H-Sachse** [reply]
Just to be clear, I assume the tasmota firmware uses the actual power value as reported on the web interface or MQTT to calculate the energy. If I take those values and integrate them over one hour, I expect power integrated over time equals energy. 
And I do not get that. I get a value that is 26% high.
Care to explain that?

**Comment by Jason2866** [comment]
Power Factor?

**Reply by H-Sachse** [reply]
Why did I even add the console logs if you don't look at them!
13:35:08.393 MQT: tasmota/tele/tasmota_158777/SENSOR = {"Time":"2024-02-03T13:35:08","ENERGY":{"TotalStartTime":"2024-01-03T15:23:55","Total":105.748,"Yesterday":13.898,"Today":4.810,"Period":49,"Power":2315,"ApparentPower":2315,"ReactivePower":0,"Factor":1.00,"Voltage":236,"Current":9.830}}

**Comment by Jason2866** [comment]
I had a look and the are of no help. Take a look at the uptime. What happens the time before? 
Btw. the wifi rssi is horrible.

**Reply by H-Sachse** [reply]
I did a firmware upgrade from 13.0.0 to 13.3.0 to see if the issue was gone, but it wasn't. I monitored before and after the upgrade and the energy calculation was just as wrong after the upgrade as before.
So how is the energy calculation done on the A1T or by Tasmota in general? Is it derived from the same power value as the one reported via web interface/MQTT? If that is the case then anything like power factor or calibration has no impact on the discrepancy I reported. Power factor or wrong calibration would already show up in the reported power value.
I'll try to provide a full hour long MQTT log the next time.

**Comment by Jason2866** [comment]
here https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xdrv_driver/xdrv_03_energy.ino 
and the device specific https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xnrg_energy/xnrg_14_bl09xx.ino#L288

**Reply by H-Sachse** [reply]
Sorry, I'm not gonna read the code now.

**Comment by H-Sachse** [comment]
Here's some logging data and a worksheet to show what I mean with the discrepancy between the power value and the accumulated energy value.
I logged my toaster with a period of 10 seconds with MQTT.
[tasmota2.log](https://github.com/arendst/Tasmota/files/14171140/tasmota2.log)
converted to csv
[tasmota2.csv](https://github.com/arendst/Tasmota/files/14171146/tasmota2.csv)
and put into a speadsheet to graph it
[tasmota2-energy-calc.ods](https://github.com/arendst/Tasmota/files/14171159/tasmota2-energy-calc.ods)
the column "expected energy" is the calculated energy according to the sampling period of 10 seconds and the reported power. 
Ignore the initial value of 12Wh, I charged my phone beforehand and didn't reset the counter. 
As can be seen, the energy reported by the A1T is higher than power * runtime curve
![grafik](https://github.com/arendst/Tasmota/assets/67382933/0fb659ed-9247-46e1-bf8f-ebccbc06e551)
The period value summed up corresponds to the reported energy, but the period value should be 2.4 for 10 seconds and 860ish watts, not 2.9

**Comment by drhuszkar** [comment]
Hi Folks!

I've found the same issue what H-Sachse described. 
The "active power" is perfect, but "energy today" and energy total" is far above the expected.
I try firmware 13, 14, 14.2 all the same. At the end of the day there is a serious difference between real comsumption and the value on nous a1t device's web interface (~30% more what is shoud be).
Oh and i try 3 different a1t plug, all the same.

I ask your advice how to solve this issue.

**Reply by arendst** [reply]
Compile the binary yourselves with the following change and report back.

Edit file ``xnrg_01_hlw8012.ino`` lines 205 to 226:
```
void HlwEverySecond(void) {
  if (Energy->data_valid[0] > ENERGY_WATCHDOG) {
    Hlw.cf1_voltage_pulse_length  = 0;
    Hlw.cf1_current_pulse_length = 0;
    Hlw.cf_power_pulse_length  = 0;
  } else {
    uint32_t hlw_len;

    if (Hlw.energy_period_counter) {

      AddLog(LOG_LEVEL_DEBUG_MORE, PSTR("HLW: EPC %u, CFlen %d usec"), Hlw.energy_period_counter, Hlw.cf_pulse_length);

      hlw_len = 10000 * 100 / Hlw.energy_period_counter;  // Add *100 to fix rounding on loads at 3.6kW (#9160)
      Hlw.energy_period_counter = 0;
      if (hlw_len) {
//        Energy->kWhtoday_delta[0] += (((Hlw.power_ratio * EnergyGetCalibration(ENERGY_POWER_CALIBRATION)) / 36) * 100) / hlw_len;
        Energy->kWhtoday_delta[0] += Energy->active_power[0] * 1000 / 36;
        EnergyUpdateToday();
      }
    }
  }
}
```

**Reply by arendst** [reply]
In addition, comparing a onetime load of 2.3kW without montoring the voltage change (which is in my case can be 30V depending on cloudiness of the sun in my street due to misbehaving PV installations) over one hour will also result in deviation of the the total 2.9kWh.

**Reply by arendst** [reply]
And ofcourse, CALIBRATE.

**Reply by H-Sachse** [reply]
Please have a look at this [post](https://github.com/arendst/Tasmota/discussions/20653#discussioncomment-8374650) of mine where I analyzed the data of one of my Nuos A1T. It clearly shows an internal discrepancy between the power values reported by the plug and the reported energy values.
I assume the energy is calculated from the same power value as is reported over MQTT.
If calibration were to effect this I would be very much surprised.

**Reply by drhuszkar** [reply]
Thank you for your fast answare.

arendst! 
You are right, the voltage changing is also the same here but the power value must be the result of current and voltage so the actual voltage shoud not be matter... 
I understand what you offer but i never compile tasmota so give me some time to understand how and i report back.
My knowledge is over with esp via esp home (homeassistant).

H-Saches!
I'm sure it is not a calibration problem, i did it as it shoud and check several times. 
My idea (yet i did not try arendst's solution) is maybe the heating of a1t cause some internal diletation. Because my observation is when i use greater power the difference become greater too.

**Reply by drhuszkar** [reply]
So, compile was much more easier than i think.
I've change what you told, the device come back online after flash and looks good. Give me a day to collect data and i report back as you ask.

**Comment by arendst** [comment]
In the meantime I dived in a bit more to find a cause of the difference.

What happens is that Tasmota Energy Monitoring calculates Apparent Power if it is not provided by the monitoring hardware (as is the case with Nous / HLW driver). Apparent Power is Voltage * Current. Active Power can never be more than Apparent Power (definition) so it adjusts measured Active Power to calculated Apparent Power.

So logging this information:
```
14:04:43.001 TEO: Power 1875.300, Delta 52133, 52093, 52091
14:04:43.003 NRG: Phase 0, kWhtoday 908, daily 0.009280000, total 21.580278397
14:04:43.326 MQT: tele/shp2/STATE = {"Time":"2024-08-19T14:04:43","Uptime":"0T00:01:21","UptimeSec":81,"Heap":26,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"POWER":"ON","Wifi":{"AP":1,"SSId":"indebuurt_IoT","BSSId":"18:E8:29:CA:17:C1","Channel":11,"Mode":"11n","RSSI":100,"Signal":-37,"LinkCount":1,"Downtime":"0T00:00:07"}}
14:04:43.339 MQT: tele/shp2/SENSOR = {"Time":"2024-08-19T14:04:43","ENERGY":{"TotalStartTime":"2023-02-24T16:08:09","Total":21.580,"Yesterday":0.001,"Today":0.009,"Period":9.08,"Power":1818.25,"ApparentPower":1818.25,"ReactivePower":0.00,"Factor":1.00,"Voltage":221.9,"Current":8.194}}
```
Shows measured Active Power (=1875.3) which is also used in energy calculation. Provided Active Power to the user is 1818.25 being the result of calculation Voltage (221.9) + Current (8.194). 

After re-calibration both measured and calculated power are equal:
```
15:02:03.001 TEO: Power 1758.300, Delta 49550, 48842, 48841
15:02:03.003 NRG: Phase 0, kWhtoday 4199, daily 0.042190000, total 21.613189697
15:02:03.341 MQT: tele/shp2/STATE = {"Time":"2024-08-19T15:02:03","Uptime":"0T00:58:41","UptimeSec":3521,"Heap":26,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":20,"MqttCount":1,"POWER":"ON","Wifi":{"AP":1,"SSId":"indebuurt_IoT","BSSId":"18:E8:29:CA:17:C1","Channel":11,"Mode":"11n","RSSI":100,"Signal":-38,"LinkCount":1,"Downtime":"0T00:00:07"}}
15:02:03.353 MQT: tele/shp2/SENSOR = {"Time":"2024-08-19T15:02:03","ENERGY":{"TotalStartTime":"2023-02-24T16:08:09","Total":21.613,"Yesterday":0.001,"Today":0.042,"Period":5.41,"Power":1758.73,"ApparentPower":1758.73,"ReactivePower":0.00,"Factor":1.00,"Voltage":217.1,"Current":8.101}}
```

Conclusion: Perform a decent calibration using a pure resistive load like a 60W bulb (at 230V). 
- The easiest way is to perform voltage calibration first using a multimeter (``Voltageset 230``). 
- Then knowing the Active power usage of the load (``PowerSet 60``). In practice your voltage will be different so you'll need to calculate Active Power accordingly.
- Finally calculate current by 60 / 230 = 261mA (``CurrentSet 261``).

This will sove your issue with pure resistive loads showing lower than measured Active Power.

**Reply by arendst** [reply]
In latest dev branch the following changes have been made:

I removed the forced correction from Apparent Power to Active Power AND Power Factor unable to be larger than 1.
This allows users to see their calibration is wrong if apparent power is shown less than active power.

Logging level 4 enables a message stating that calibration is needed when active power is more than 0.005% from apparent power.

The changes to the code will solve the mismatch between user calculated power and Tasmota calculated Energy usage. The mismatch is caused by bad calibration between Power/Voltage and Current.

**Comment by drhuszkar** [comment]
Thank you for your investigation. 
I'm sorry to cause more work beacuse of my fault. I did the calibration as best i coud. The voltage here is changing so fast, it look impossible to calculate power. 

In my practice i use two multimeter
First i set the voltage from multimeter
Second to set current from multimeter
And finally i set power if necessary, usually it is not because it is voltage * current and fine.

Before i write my issue i recalibrate some a1t and test them but i'am going to check them again.

What does it mean 'dev branch'? Is it a beta firmware?

(I do not forget what i promise, yesterday my wife unplug my electric heater)

**Reply by arendst** [reply, marked as answer]
The dev branch is the development branch of Tasmota with latest and greatest features and can be found here https://ota.tasmota.com/tasmota/

Today I've added a new command allowing to calibrate a lot easier.

The new procedure is as follows:
- Calibrate Voltage using a multimeter and command ``VoltageSet <voltage read from multimeter>``
- Connect a known resistive load like a 60W bulb. On the bulb you'll find a remark that this 60W is valid if exactly 230V is provided. The new command ``PowerSet 60,230`` will use the now calibrated Voltage and calculate the correct power and current calibration values even if voltage is fluctuating at the time you execute the command.

That's it! No seperate ``PowerSet`` and ``CurrentSet`` needed.

The main issue here is you'll need to know from you resistive load the exact Voltage at which the Power will be provided. A Heater is not a good device for calibration as it changes it's load depending on the heat.

**Reply by drhuszkar** [reply]
Thank you very much for your fast support!

I'm going to try today evening with 14.2.0.1 when voltage is bit more stabel here. And of course i write my experience.

I hope you do not mind if ask something offtopic.
When i first buy from mediarath .com, there was an option to support tasmota. I send some support but i get no back any comfirmation so i was not sure this money reach you. My question, when i use this option the help arrive to its destination?

**Reply by arendst** [reply]
Yes. From time to time I receive tips from Mediarath so yours must have been with it. Thx!

**Reply by drhuszkar** [reply]
Great, next order i kepp this practice.

I never like cloud solutions, and not use them ever. When i found tasmota it has changed my life. It is not a joke or exaggeration. Tasmota drive me to homeassistant and now i have my custom smart solutions with several device which save me money, give me more comfort, beacuse of you. So thank you again!

**Reply by drhuszkar** [reply]
I can not wait and have an issue with my 60w bulb

voltage set made really carefully
current and voltage set to default
237 V
0.281 A (on my multimeter there is 0.257)
CMD: CurrentSet 60,230
MQT: pwr_17_stat/pwr_17_155BBB/RESULT = {"CurrentSetCal":746}

now i see 0.060 A wich means it not care the ',230' value in my order

of course firmware is 14.2.0.1

**Reply by dia-de-muertos** [reply]
> The dev branch is the development branch of Tasmota with latest and greatest features and can be found here https://ota.tasmota.com/tasmota/
> 
> Today I've added a new command allowing to calibrate a lot easier.
> 
> The new procedure is as follows:
> 
>     * Calibrate Voltage using a multimeter and command `VoltageSet <voltage read from multimeter>`
> 
>     * Connect a known resistive load like a 60W bulb. On the bulb you'll find a remark that this 60W is valid if exactly 230V is provided. The new command `PowerSet 60,230` will use the now calibrated Voltage and calculate the correct power and current calibration values even if voltage is fluctuating at the time you execute the command.
> 
> 
> That's it! No seperate `PowerSet` and `CurrentSet` needed.
> 

If I measure and set a voltage of 227V and connect a bulb that is marked with 100W@230V,
I also have to use 'PowerSet 100,230' and NOT 'PowerSet 100,227'. Am I right?

**Reply by Jason2866** [reply]
The bulb has **only** 100W when powered with 230V. So yes you have to use this setting. Tasmota calibration routine calculates the correct values for 227 out of this.

**Comment by drhuszkar** [comment]
After i re calibrate manually as before i see powerfactor 1.01 as you told

17:04:37.403 MQT: pwr_17_tele/pwr_17_155BBB/SENSOR = {"Time":"2024-08-20T17:04:37","ENERGY":{"TotalStartTime":"2024-08-15T09:41:04","Total":0.335,"Yesterday":0.000,"Today":0.059,"Power":61,"ApparentPower":61,"ReactivePower":3277,"Factor":1.00,"Voltage":237,"Current":0.256}}
17:04:39.268 MQT: pwr_17_tele/pwr_17_155BBB/SENSOR = {"Time":"2024-08-20T17:04:39","ENERGY":{"TotalStartTime":"2024-08-15T09:41:04","Total":0.335,"Yesterday":0.000,"Today":0.059,"Power":61,"ApparentPower":61,"ReactivePower":3277,"Factor":1.01,"Voltage":236,"Current":0.256}}

**Reply by arendst** [reply]
OK.

You seem to have the version just before the latest ;-)

Pls redownload the binary dated 20240820 14:17 and try again.

**Reply by drhuszkar** [reply]
Yeh, thats work

Now i see on my multimeter 0.257 A, on tasmota 0.269 A and 64W
It is a 60 EUR instrument so i can believe it is not as exact as to be
Tomorrow a make an other test and report back the difference between an "old" a1t and this new.

Thank you!

**Reply by drhuszkar** [reply]
I wait the evening to recalibrate all my a1t but i looks like you solved my problem.
I have a 600w heater (ye i know it is not the best but biggest load right now what i can swiching as i need). A1t calibrated with 60w bulb with the heater shows 599w max and 2 hours continous heating 1.208kw.

Can i use PowerSet 60,230 command for a 100w bulb with PowerSet 100,230 ?
Is it true calibrating with biggest comsumption device is better?

**Reply by arendst** [reply]
> Can i use PowerSet 60,230 command for a 100w bulb with PowerSet 100,230

Yes as long as the bulb states it draws 100W at 230V

> Is it true calibrating with biggest comsumption device is better

I noticed with the latest command there is no difference between calibrating with a 60W or higher load so your 100W is just fine.

**Reply by drhuszkar** [reply]
OK, it looks like everything is fine. The daily consumption looks real.
I upgrade and recalibrate all my a1t, tomorrow i see.

One interesting thing, after voltageset i send PowerSet 60,230 and this sometime changes the voltage reading. Maybe it was a jump in the grid while calibrating voltage, i do not know. Just it was intereting.

So thank you again, i hope tomorrow everything will be fine.

**Reply by drhuszkar** [reply]
One day pass with more a1t recalibrated with 14.2.0.1 fw and i have enough experience to share my opinion.

Arendst! With your help, the new method of calibration i can finally set up my utility meters as i needed.
Thank you for your work!

**Comment by drhuszkar** [comment]
One important note if enyone comes here with utility meter calibration problem later!

**DO NOT CALIBRATE IF YOU HAVE PV SYSTEM AND SUN IS SHINING !!!**

Your voltage will dancing (you can not catch a stabil point for calibration) and there will be more difference between your swich box (where your inverter connect to the grid) and between your electric plugs inside the house.

**Comment by H-Sachse** [comment]
With the updated firmware and the new calibration procedure the energy values are now OK for my EV charging.
@arendst: Thanks for the fix!

**Comment by beneix** [comment]
I have just purchased a Nous A1T and calibrated it using a Fluke true RMS multimeter. After calibration, there is something weird with the measurements: a 60W incandescent bulb gives:
Active Power 59.8W
Apparent Power 78.96W
Reactive Power 51.8W
Power Factor 0.76

After I calibrated the voltage but before calibrating the power, it was showing the same current but 76W active power, 78.66 apparent power, 20.3 reactive power and a power factor of 0.97.

This is clearly off. I am running 15.4.0. Any ideas?
